Sources/SourceKitLSP/Swift/SyntaxHighlightingTokenParser.swift

```diff
@@ -0,0 +1,234 @@
+struct SyntaxHighlightingTokenParser {
+  private let sourcekitd: SourceKitD
+  /// Use `keys.nameoffset` and `keys.namelength` instead of `keys.offset`
+  /// and `keys.length`. This is useful in AST structures provided by
+  /// sourcekitd, since these contain both a name range and a normal range,
+  /// where the latter refers to the entire node (e.g. the entire function
+  /// declaration).
+  private let useName: Bool
+
+  init(sourcekitd: SourceKitD, useName: Bool = false) {
+    self.sourcekitd = sourcekitd
+    self.useName = useName
+  }
+
+  func parseTokens(_ response: SKDResponseDictionary, in snapshot: DocumentSnapshot) -> [SyntaxHighlightingToken] {
+    let keys = sourcekitd.keys
+    var tokens: [SyntaxHighlightingToken] = []
+
+    if let offset: Int = useName ? response[keys.nameoffset] : response[keys.offset],
+       var length: Int = useName ? response[keys.namelength] : response[keys.length],
+       let start: Position = snapshot.positionOf(utf8Offset: offset),
+       let skKind: sourcekitd_uid_t = response[keys.kind],
+       let (kind, modifiers) = parseKindAndModifiers(skKind) {
+
+      // We treat function declaration and enum member name tokens as a special
+      // case, e.g. SourceKit returns `f(x: Int, y: Int)` as a name instead of just `f`.
+      if useName && [.function, .method, .enumMember].contains(kind) && modifiers.contains(.declaration),
+         let name: String = response[keys.name],
+         name.contains("("),
+         let funcNameLength: Int = name.split(separator: "(").first?.utf8.count {
+        length = funcNameLength
+      }
+
+      // If the name is escaped in backticks, we need to add two characters to the
+      // length for the backticks.
+      if modifiers.contains(.declaration),
+         let index = snapshot.indexOf(utf8Offset: offset), snapshot.text[index] == "`" {
+        length += 2
+      }
+
+      if let end: Position = snapshot.positionOf(utf8Offset: offset + length) {
+        let multiLineRange = start..<end
+        let ranges = multiLineRange.splitToSingleLineRanges(in: snapshot)
+
+        tokens += ranges.map {
+          SyntaxHighlightingToken(
+            range: $0,
+            kind: kind,
+            modifiers: modifiers
+          )
+        }
+      }
+    }
+
+    if let substructure: SKDResponseArray = response[keys.substructure] {
+      tokens += parseTokens(substructure, in: snapshot)
+    }
+
+    return tokens
+  }
+
+  func parseTokens(_ response: SKDResponseArray, in snapshot: DocumentSnapshot) -> [SyntaxHighlightingToken] {
+    var result: [SyntaxHighlightingToken] = []
+    response.forEach { (_, value) in
+      result += parseTokens(value, in: snapshot)
+      return true
+    }
+    return result
+  }
+
+  private func parseKindAndModifiers(_ uid: sourcekitd_uid_t) -> (SyntaxHighlightingToken.Kind, SyntaxHighlightingToken.Modifiers)? {
+    let api = sourcekitd.api
+    let values = sourcekitd.values
+    switch uid {
```

Comment:
Interestingly yes, there seems to be `key.is_system`, which we can use for that. (Though I am not sure whether UIKit counts as system module) Thanks!

Reply:
Interesting. The meaning of the `defaultLibrary` modifier is not documented at all in LSP. In VSCode, it's documented as "For symbols that are part of the standard library".  I guess it depends on how broad a definition of "standard library" we want to use. When describing it in words, we usually distinguish the standard library (e.g. Swift module) from the broader Apple SDK (e.g. UIKit). Both modules are considered "system", which includes the stdlib, the SDK and also any system search paths provided to the compiler (e.g. via `-Fsystem`).

For things like symbol colouring we've always used "is_system" in the past, so I'm happy to treat all system library symbols as `defaultLibrary` for now unless we get further clarification in the spec that makes us change our mind.

Reply:
Yeah, I added similar support to clangd in https://reviews.llvm.org/D101554, I think this behavior is fine - I'm not sure of any precedent from other languages like Java/Python. `defaultLibrary` seems to make sense since it's the closest modifier in the spec (and therefore themes can easily support it).